File: DSA/neetcode150/src/py/largest_rectangle_in_histogram.py
class Solution:
    def largestRectangleArea(self, heights: list[int]) -> int:
        n = len(heights)
        lp = [-1]*n
        stk = []
        for i in range(n):
            while stk and heights[stk[-1]] >= heights[i]:
                stk.pop()
            if stk:
                lp[i] = stk[-1]
            stk.append(i)

        rp = [n]*n
        stk = []
        for i in range(n-1, -1, -1):
            while stk and heights[i] <= heights[stk[-1]]:
                stk.pop()
            if stk:
                rp[i] = stk[-1]
            stk.append(i)

        result = 0
        for i, height in enumerate(heights):
            area = height * (rp[i] - lp[i] + 1)
            result = max(result, area)

        return result

File: DSA/src/go/largest_rectangle_in_histogram.go
package main

func getIntArray(size int, def int) []int {
	arr := make([]int, size)
	for i := range arr {
		arr[i] = def
	}
	return arr
}

func largestRectangleArea(heights []int) int {
	n := len(heights)
	left := getIntArray(n, -1)
	var stk []int
	for i := 0; i < n; i++ {
		for len(stk) != 0 && heights[i] <= heights[stk[len(stk)-1]] {
			stk = stk[:len(stk)-1]
		}
		if len(stk) != 0 {
			left[i] = stk[len(stk)-1]
		}
		stk = append(stk, i)
	}

	right := getIntArray(n, n)
	stk = []int{}
	for i := n - 1; i >= 0; i-- {
		for len(stk) != 0 && heights[i] <= heights[stk[len(stk)-1]] {
			stk = stk[:len(stk)-1]
		}
		if len(stk) != 0 {
			right[i] = stk[len(stk)-1]
		}
		stk = append(stk, i)
	}

	var result, area int
	for i := range heights {
		area = heights[i] * (right[i] - left[i] - 1)
		result = max(result, area)
	}
	return result
}
